src/common/helpers/application-lock.js

```diff
@@ -0,0 +1,188 @@
+      grantVersion,
+      sbi,
+      ownerId,
+      lockedAt: now,
+      expiresAt
+    })
+
+    return {
+      _id: insertResult.insertedId,
+      grantCode,
+      grantVersion,
+      sbi,
+      ownerId,
+      lockedAt: now,
+      expiresAt
+    }
+  } catch (err) {
+    if (err.code === 11000) {
+      return null
+    }
+    throw err
+  }
+}
+
+/**
+ * Acquires an exclusive lock for an application.
+ *
+ * Lock acquisition rules:
+ *  - Only one user may hold a lock for a given application at a time
+ *  - Expired locks may be taken over
+ *  - The same user may re-acquire (refresh) their own lock
+ *  - If another active user holds the lock, null is returned
+ *
+ * @param {import('mongodb').Db} db - MongoDB database instance
+ * @param {Object} params
+ * @param {string} params.grantCode
+ * @param {number} params.grantVersion
+ * @param {string} params.sbi
+ * @param {string} params.ownerId - DefraID user ID
+ * @returns {Promise<Object|null>} Lock document if acquired, otherwise null
+ */
+export async function acquireApplicationLock(db, { grantCode, grantVersion, sbi, ownerId }) {
+  const now = new Date()
+  const expiresAt = new Date(now.getTime() + LOCK_TTL_MS)
+  const collection = db.collection('application-locks')
+
+  const existingLock = await tryAcquireExistingLock(
+    collection,
+    { grantCode, grantVersion, sbi, ownerId },
+    { now, expiresAt }
+  )
+
+  if (existingLock) {
+    return existingLock
+  }
+
+  return createNewLock(collection, { grantCode, grantVersion, sbi, ownerId }, { now, expiresAt })
+}
+
+/**
+ * Extends the expiry time of an existing application lock.
+ *
+ * The lock will only be refreshed if it is owned by the given user.
+ *
+ * @param {import('mongodb').Db} db - MongoDB database instance
+ * @param {Object} params
+ * @param {string} params.grantCode
+ * @param {number} params.grantVersion
+ * @param {string} params.sbi
+ * @param {string} params.ownerId - DefraID user ID
+ * @returns {Promise<boolean>} True if the lock was refreshed, false otherwise
+ */
+export async function refreshApplicationLock(db, { grantCode, grantVersion, sbi, ownerId }) {
+  const now = new Date()
+  const expiresAt = new Date(now.getTime() + LOCK_TTL_MS)
+
+  const result = await db.collection('application-locks').updateOne(
+    {
+      grantCode,
+      grantVersion,
+      sbi,
+      ownerId
+    },
+    {
+      $set: { expiresAt }
+    }
+  )
+
+  return result.matchedCount === 1
+}
+
+/**
+ * Releases an application lock held by the given user.
+ *
+ * If no matching lock exists, this operation is a no-op.
+ *
+ * @param {import('mongodb').Db} db - MongoDB database instance
+ * @param {Object} params
+ * @param {string} params.grantCode
+ * @param {number} params.grantVersion
+ * @param {string} params.sbi
+ * @param {string} params.ownerId - DefraID user ID
+ * @returns {Promise<void>}
+ */
+export async function releaseApplicationLock(db, { grantCode, grantVersion, sbi, ownerId }) {
```

Comment:
I know we only call this function, but could it return a 1 for success and 0 for fail? Seems better long term maintaince.